Azure Arc enabled data services on Microk8s: DC, SQL MI and PostgreSQL scenario

mdrakiburrahman opened the pull request:
[Issue Reference](https://github.com/microsoft/azure_arc/issues/549) - this PR contains the scenario for Vanilla DC, SQL MI & Postgres deployment on Microk8s.

Used the latest [cluster_api](https://github.com/microsoft/azure_arc/tree/main/azure_arc_data_jumpstart/cluster_api/capi_azure/arm_template) scenario as a baseline (where SQL MI and PostgreSQL is deployed via booleans)- with few notable changes/slight improvements:
* Decoupled VNET deployment into new ARM template - `VNET.json`. This fixed the occasional race condition (reproduced twice) issue I was facing where Client VM's NIC would start deploying faster than the Ubuntu ARM template could deploy the VNET
* made `templateBaseUrl` a parameter that gets passed in with `azuredeploy.parameters.json`. This parameter is then used throughout the ARM templates to reference the github root path for scripts that are pulled as part of several VM Extensions run - such as `Bootstrap.ps1`, `install_microk8s.sh` etc.

  Updated `azuredeploy.parameters.example.json` with default (root path of this repo) and Readme with how to use `templateBaseUrl` during deployment.

* New `install_microk8s.sh` includes Microk8s install and some post-deployment config to prep for Arc
* Some cosmetic upgrades/added comments to `DataServicesLogonScript.ps1`
* Data Controller, SQL MI and Hyperscale ARM parameters in updated to work with Microk8s (New Storage Class and NodePort)
* Scenario documentation/arch diagrams/screenshots/gif etc.

likamrat commented on azure_arc_data_jumpstart/microk8s/azure/arm_template/VNET.json:
```diff
@@ -0,0 +1,54 @@
+  "resources": [
+    {
+      "type": "Microsoft.Network/virtualNetworks",
+      "comments": "Deploys a VNET and Subnet for Sidecar and K8s VM",
```
```suggestion
      "comments": "Deploys a VNET and Subnet for Client and K8s VM",
```

We are keeping the "Client" wording for now and we will re-visit the sidecar terminology in the future. Make sure to update wherever needed as well as screenshots.

mdrakiburrahman replied:
Agreed, updated markdown and Screenshot.

likamrat commented on azure_arc_data_jumpstart/microk8s/azure/arm_template/artifacts/Bootstrap.ps1:
```diff
@@ -0,0 +1,142 @@
+Write-Host "Extending C:\ partition to the maximum size"
+Resize-Partition -DriveLetter C -Size $(Get-PartitionSupportedSize -DriveLetter C).SizeMax
+
+# Downloading Github artifacts for DataServicesLogonScript.ps1
```
```suggestion
# Downloading GitHub artifacts for DataServicesLogonScript.ps1
```

likamrat commented on azure_arc_data_jumpstart/microk8s/azure/arm_template/artifacts/DeployPostgreSQL.ps1:
```diff
@@ -0,0 +1,76 @@
+Write-Host "Azure Arc PostgreSQL Hyperscale is ready!"
+Write-Host "`n"
+
+
```
```suggestion
Start-Sleep -Seconds 60

```

This is required to make sure PG is up and running completely and so the database restore will not fail. See [here](https://github.com/microsoft/azure_arc/blob/main/azure_arc_data_jumpstart/cluster_api/capi_azure/arm_template/artifacts/DeployPostgreSQL.ps1#L45).

likamrat commented on azure_arc_data_jumpstart/microk8s/azure/arm_template/ubuntuMicrok8s.json:
```diff
@@ -0,0 +1,260 @@
+    "resourceTags": {
+      "type": "object",
+      "defaultValue": {
+        "Project": "jumpstart_azure_arc_data"
```
```suggestion
        "Project": "jumpstart_azure_arc_data_services"
```

Make sure this is the tag across the board wherever tags are declared

likamrat commented on docs/azure_arc_jumpstart/azure_arc_data/microk8s/_index.md:
```diff
@@ -0,0 +1,8 @@
+type: docs
+title: "MicroK8s"
+linkTitle: "MicroK8s"
+weight: 4
```
```suggestion
weight: 5
```

Since this comes before the "Upstream Kubernetes (Kubeadm)" section in the Jumpstart, also update this section with "_weight: 6_"

mdrakiburrahman replied:
Ah I thought this had to do with font weight, didn't realize [this](https://bwaycer.github.io/hugo_tutorial.hugo/content/ordering/) is how Hugo ordering works.

Updated microk8s, as well as kubeadm.

likamrat commented on docs/azure_arc_jumpstart/azure_arc_data/microk8s/azure/arm_template/dc_vanilla/_index.md:
````diff
@@ -0,0 +1,206 @@
+
+  > **Note: It is optional, but highly recommended, to scope the SP to a specific [Azure subscription and resource group](https://docs.microsoft.com/en-us/cli/azure/ad/sp?view=azure-cli-latest).**
+
+- Enable subscription for the _Microsoft.AzureArcData_ resource provider for Azure Arc enabled data services. Registration is an asynchronous process, and registration may take approximately 10 minutes.
+
+  ```shell
+  az provider register --namespace Microsoft.AzureArcData
+  ```
+
+  You can monitor the registration process with the following commands:
+
+  ```shell
+  az provider show -n Microsoft.AzureArcData -o table
+  ```
````
This should go inside to automation. See [here](https://github.com/microsoft/azure_arc/blob/main/azure_arc_data_jumpstart/cluster_api/capi_azure/arm_template/artifacts/DataServicesLogonScript.ps1#L38)

mdrakiburrahman replied:
Removed from markdown, as it's already inside [`DataServicesLogonScript.ps1`](https://github.com/microsoft/azure_arc/blob/60497dd3f3b590b97c2ca52b9f6849e989d9c21e/azure_arc_data_jumpstart/microk8s/azure/arm_template/artifacts/DataServicesLogonScript.ps1) to your point.

Only point I want to highlight is, `az provider register` will require the Service Principal the user creates to have **Subscription** level [**Owner**/**Contributor**](https://docs.microsoft.com/en-us/azure/azure-resource-manager/management/resource-providers-and-types#register-resource-provider) permissions to register the provider.

So for example, say someone runs this scenario the very first time without having already registering `AzureArcData` before, and the SP they created is only RG scoped - in this situation when the script runs inside the Client VM, the SP won't be able to register the provider.

(But if this happens in this scenario, the user can run the registrations manually using their own creds against the Subscription).

likamrat commented on docs/azure_arc_jumpstart/azure_arc_data/microk8s/azure/arm_template/dc_vanilla/_index.md:
```diff
@@ -0,0 +1,206 @@
+
+From the [Microk8s Github repo](https://github.com/ubuntu/microk8s):
+
+_"Microk8s is a single-package, fully conformant, lightweight Kubernetes that works on 42 flavours of Linux. Perfect for Developer workstations, IoT, Edge & CI/CD. MicroK8s tracks upstream and releases beta, RC and final bits the same day as upstream K8s."_
+
+In this guide, we automate the installation of Microk8s on an Ubuntu 18.04 VM running on Azure using a few simple commands to install from the [Snap Store](https://snapcraft.io/microk8s), before proceeding to onboard it as an Azure Arc enabled Kubernetes Cluster.
+
+Once our K8s Cluster is onboarded, we proceed to create a [Custom Location](https://docs.microsoft.com/en-us/azure/azure-arc/kubernetes/custom-locations), and deploy an Azure Arc Data Controller in [Directly Connected mode](https://docs.microsoft.com/en-us/azure/azure-arc/data/connectivity#connectivity-modes).
```
```suggestion
_"Microk8s is a single-package, fully conformant, lightweight Kubernetes that works on 42 flavors of Linux. Perfect for Developer workstations, IoT, Edge & CI/CD. MicroK8s tracks upstream and releases beta, RC and final bits the same day as upstream K8s."_

In this guide, we automate the installation of Microk8s on an Ubuntu 18.04 VM running on Azure using a few simple commands to install from the [Snap Store](https://snapcraft.io/microk8s), before proceeding to onboard it as an Azure Arc enabled Kubernetes Cluster.

Once our K8s Cluster is onboarded, we proceed to create a [Custom Location](https://docs.microsoft.com/en-us/azure/azure-arc/kubernetes/custom-locations), and deploy an Azure Arc Data Controller in [Directly Connected mode](https://docs.microsoft.com/en-us/azure/azure-arc/data/connectivity#connectivity-modes).
```

likamrat commented on docs/azure_arc_jumpstart/azure_arc_data/microk8s/azure/arm_template/dc_vanilla/_index.md:
```diff
@@ -0,0 +1,206 @@
+  - `logAnalyticsWorkspaceName` - Unique name for log analytics workspace deployment.
+  - `deploySQLMI` - Boolean that sets whether or not to deploy SQL Managed Instance, for this data controller only scenario we leave it set to _**false**_.
+  - `deployPostgreSQL` - Boolean that sets whether or not to deploy PostgreSQL Hyperscale, for this data controller only scenario we leave it set to _**false**_.
+  - `templateBaseUrl` - Github URL to the deployment template - filled in by default to point to [Microsoft/Azure Arc](https://github.com/microsoft/azure_arc) repository, but you can point this to your forked repo as well.
```
```suggestion
  - `templateBaseUrl` - GitHub URL to the deployment template - filled in by default to point to [Microsoft/Azure Arc](https://github.com/microsoft/azure_arc) repository, but you can point this to your forked repo as well.
```

likamrat commented on docs/azure_arc_jumpstart/azure_arc_data/microk8s/azure/arm_template/dc_vanilla/_index.md:
```diff
@@ -0,0 +1,206 @@
+
+  ![Deployment time](./02.png)
+
+- Once Azure resources has been provisioned, you will be able to see it in the Azure portal. At this point, the resource group should have **13** various Azure resources deployed.
```
```suggestion
- Once Azure resources have been provisioned, you will be able to see it in the Azure portal. At this point, the resource group should have **13 various Azure resources deployed**.
```

likamrat commented on docs/azure_arc_jumpstart/azure_arc_data/microk8s/azure/arm_template/mssql_mi/_index.md:
```diff
@@ -0,0 +1,212 @@
+
+  ![Deployment time](./02.png)
+
+- Once Azure resources has been provisioned, you will be able to see it in the Azure portal. At this point, the resource group should have **13** various Azure resources deployed.
```
```suggestion
- Once Azure resources have been provisioned, you will be able to see it in the Azure portal. At this point, the resource group should have **13 various Azure resources deployed**.
```

likamrat commented on docs/azure_arc_jumpstart/azure_arc_data/microk8s/azure/arm_template/mssql_mi/_index.md:
```diff
@@ -0,0 +1,212 @@
+
+  ![Wallpaper Change](./07.png)
+
+- Since this scenario is deploying the Azure Arc Data Controller and SQL Managed Instance, you will also notice additional newly deployed Azure resources in the resources group (at this point you should have **17** various Azure resources deployed). The important ones to notice are:
```
```suggestion
- Since this scenario is deploying the Azure Arc Data Controller and SQL Managed Instance, you will also notice additional newly deployed Azure resources in the resources group (at this point you should have **17 various Azure resources deployed**. The important ones to notice are:
```

likamrat commented on docs/azure_arc_jumpstart/azure_arc_data/microk8s/azure/arm_template/postgresql_hyperscale/_index.md:
```diff
@@ -0,0 +1,212 @@
+
+  ![Deployment time](./02.png)
+
+- Once Azure resources has been provisioned, you will be able to see it in the Azure portal. At this point, the resource group should have **13** various Azure resources deployed.
```
```suggestion
- Once Azure resources have been provisioned, you will be able to see it in the Azure portal. At this point, the resource group should have **13 various Azure resources deployed**.
```

likamrat commented on docs/azure_arc_jumpstart/azure_arc_data/microk8s/azure/arm_template/postgresql_hyperscale/_index.md:
```diff
@@ -0,0 +1,212 @@
+
+  ![Wallpaper Change](./07.png)
+
+- Since this scenario is deploying the Azure Arc Data Controller and PostgreSQL Hyperscale instance, you will also notice additional newly deployed Azure resources in the resources group (at this point you should have **17** various Azure resources deployed). The important ones to notice are:
```
```suggestion
- Since this scenario is deploying the Azure Arc Data Controller and PostgreSQL Hyperscale instance, you will also notice additional newly deployed Azure resources in the resources group (at this point you should have **17 various Azure resources deployed**. The important ones to notice are:
```

likamrat commented on docs/azure_arc_jumpstart/azure_arc_data/microk8s/azure/arm_template/dc_vanilla/_index.md:
```diff
@@ -0,0 +1,206 @@
+
+- Let the script run it's course and **do not close** the PowerShell session, this will be done for you once completed.
+
+  ![PowerShell logon script run](./06.gif)
```
technically, this should be "_01.gif_"

mdrakiburrahman replied:
Agreed, updated gif name, as well as decremented pngs and updated markdown.

likamrat commented on docs/azure_arc_jumpstart/azure_arc_data/microk8s/azure/arm_template/mssql_mi/_index.md:
```diff
@@ -0,0 +1,212 @@
+
+- Now that first phase of the automation is completed, it is time to RDP to the Client VM using it's public IP.
+
+  ![Client VM public IP](./05.png)
```
See comment on "Sidecar" > "Client"

mdrakiburrahman replied:
Updated screenshot

likamrat commented on docs/azure_arc_jumpstart/azure_arc_data/microk8s/azure/arm_template/mssql_mi/_index.md:
````diff
@@ -0,0 +1,212 @@
+
+  > **Note: It is optional, but highly recommended, to scope the SP to a specific [Azure subscription and resource group](https://docs.microsoft.com/en-us/cli/azure/ad/sp?view=azure-cli-latest).**
+
+- Enable subscription for the _Microsoft.AzureArcData_ resource provider for Azure Arc enabled data services. Registration is an asynchronous process, and registration may take approximately 10 minutes.
+
+  ```shell
+  az provider register --namespace Microsoft.AzureArcData
+  ```
+
+  You can monitor the registration process with the following commands:
+
+  ```shell
+  az provider show -n Microsoft.AzureArcData -o table
+  ```
````
See previous comment on provider registration

mdrakiburrahman replied:
Agreed, removed from markdown. Commented earlier on possible edge case FYI.

likamrat commented on docs/azure_arc_jumpstart/azure_arc_data/microk8s/azure/arm_template/mssql_mi/_index.md:
```diff
@@ -0,0 +1,212 @@
+
+## Architecture (In a nutshell)
+
+From the [Microk8s Github repo](https://github.com/ubuntu/microk8s):
+
+_"Microk8s is a single-package, fully conformant, lightweight Kubernetes that works on 42 flavours of Linux. Perfect for Developer workstations, IoT, Edge & CI/CD. MicroK8s tracks upstream and releases beta, RC and final bits the same day as upstream K8s."_
+
+In this guide, we automate the installation of Microk8s on an Ubuntu 18.04 VM running on Azure using a few simple commands to install from the [Snap Store](https://snapcraft.io/microk8s), before proceeding to onboard it as an Azure Arc enabled Kubernetes Cluster.
+
+Once our K8s Cluster is onboarded, we proceed to create a [Custom Location](https://docs.microsoft.com/en-us/azure/azure-arc/kubernetes/custom-locations), and deploy an Azure Arc Data Controller in [Directly Connected mode](https://docs.microsoft.com/en-us/azure/azure-arc/data/connectivity#connectivity-modes).
```
```suggestion
From the [Microk8s Github repo](https://github.com/ubuntu/microk8s):

_"Microk8s is a single-package, fully conformant, lightweight Kubernetes that works on 42 flavors of Linux. Perfect for Developer workstations, IoT, Edge & CI/CD. MicroK8s tracks upstream and releases beta, RC and final bits the same day as upstream K8s."_

In this guide, we automate the installation of Microk8s on an Ubuntu 18.04 VM running on Azure using a few simple commands to install from the [Snap Store](https://snapcraft.io/microk8s), before proceeding to onboard it as an Azure Arc enabled Kubernetes Cluster.

Once our K8s Cluster is onboarded, we proceed to create a [Custom Location](https://docs.microsoft.com/en-us/azure/azure-arc/kubernetes/custom-locations), and deploy an Azure Arc Data Controller in [Directly Connected mode](https://docs.microsoft.com/en-us/azure/azure-arc/data/connectivity#connectivity-modes).
```

likamrat commented on docs/azure_arc_jumpstart/azure_arc_data/microk8s/azure/arm_template/postgresql_hyperscale/_index.md:
````diff
@@ -0,0 +1,212 @@
+
+  > **Note: It is optional, but highly recommended, to scope the SP to a specific [Azure subscription and resource group](https://docs.microsoft.com/en-us/cli/azure/ad/sp?view=azure-cli-latest).**
+
+- Enable subscription for the _Microsoft.AzureArcData_ resource provider for Azure Arc enabled data services. Registration is an asynchronous process, and registration may take approximately 10 minutes.
+
+  ```shell
+  az provider register --namespace Microsoft.AzureArcData
+  ```
+
+  You can monitor the registration process with the following commands:
+
+  ```shell
+  az provider show -n Microsoft.AzureArcData -o table
+  ```
````
See the previous comment on provider registration

mdrakiburrahman replied:
Agreed, removed from markdown. Commented earlier on possible edge case FYI.

likamrat commented on docs/azure_arc_jumpstart/azure_arc_data/microk8s/azure/arm_template/postgresql_hyperscale/_index.md:
```diff
@@ -0,0 +1,212 @@
+
+From the [Microk8s Github repo](https://github.com/ubuntu/microk8s):
+
+_"Microk8s is a single-package, fully conformant, lightweight Kubernetes that works on 42 flavours of Linux. Perfect for Developer workstations, IoT, Edge & CI/CD. MicroK8s tracks upstream and releases beta, RC and final bits the same day as upstream K8s."_
+
+In this guide, we automate the installation of Microk8s on an Ubuntu 18.04 VM running on Azure using a few simple commands to install from the [Snap Store](https://snapcraft.io/microk8s), before proceeding to onboard it as an Azure Arc enabled Kubernetes Cluster.
+
+Once our K8s Cluster is onboarded, we proceed to create a [Custom Location](https://docs.microsoft.com/en-us/azure/azure-arc/kubernetes/custom-locations), and deploy an Azure Arc Data Controller in [Directly Connected mode](https://docs.microsoft.com/en-us/azure/azure-arc/data/connectivity#connectivity-modes).
```
```suggestion
_"Microk8s is a single-package, fully conformant, lightweight Kubernetes that works on 42 flavors of Linux. Perfect for Developer workstations, IoT, Edge & CI/CD. MicroK8s tracks upstream and releases beta, RC and final bits the same day as upstream K8s."_

In this guide, we automate the installation of Microk8s on an Ubuntu 18.04 VM running on Azure using a few simple commands to install from the [Snap Store](https://snapcraft.io/microk8s), before proceeding to onboard it as an Azure Arc enabled Kubernetes Cluster.

Once our K8s Cluster is onboarded, we proceed to create a [Custom Location](https://docs.microsoft.com/en-us/azure/azure-arc/kubernetes/custom-locations), and deploy an Azure Arc Data Controller in [Directly Connected mode](https://docs.microsoft.com/en-us/azure/azure-arc/data/connectivity#connectivity-modes).
```

likamrat reviewed (changes requested):
* "Sidecar" > "Client"
* updated tags
* Embed the AzureArcData provider registration in the automation
* Great addition with the diagram and the gif!
* Minor lingo and syntax comments

mdrakiburrahman commented:
@likamrat commits [`3bc95ca`](https://github.com/microsoft/azure_arc/pull/620/commits/3bc95cadd6c9db9dc98d8f269c7d8df13a312434) and [`60497dd`](https://github.com/microsoft/azure_arc/pull/620/commits/60497dd3f3b590b97c2ca52b9f6849e989d9c21e) capture the following fixes:

- Ensure no "Sidecar" references (including Screenshots)
- Included `Start-Sleep` for Postgres before running restore
- Updated all tags to `jumpstart_azure_arc_data_services`
- Moved hugo ordering: Microk8s:5, Kubeadm:6
- Removed `Microsoft.AzureArcData` as it's already in Automation
- Moved gif to `01.gif`
- Decremented pngs due to gif being moved to it's own order, updated markdown

Commented inline as well.

likamrat reviewed (approved):
LGTM. Great work @mdrakiburrahman!!
